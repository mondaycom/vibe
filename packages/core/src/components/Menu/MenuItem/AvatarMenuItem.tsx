import React, { ForwardedRef, forwardRef, useCallback } from "react";
import MenuItem, { MenuItemProps } from "./MenuItem";
import Avatar, { AvatarProps } from "../../Avatar/Avatar";
import VibeComponentProps from "../../../types/VibeComponentProps";

export interface AvatarMenuItemProps extends VibeComponentProps {
  /**
   * Props passed to the avatar component.
   */
  avatarProps?: AvatarProps;
  /**
   * Props passed to the menu item component.
   */
  menuItemProps?: MenuItemProps;
}

/**
 * MenuItem with Avatar instead of Icon
 */
const AvatarMenuItem: React.FC<AvatarMenuItemProps & { isMenuChild?: boolean; isSelectable?: boolean }> = forwardRef(
  ({ avatarProps, menuItemProps, ...embeddedMenuItemProps }, ref: ForwardedRef<HTMLElement>) => {
    const renderAvatar = useCallback(() => <Avatar {...avatarProps} />, [avatarProps]);
    return <MenuItem {...embeddedMenuItemProps} {...menuItemProps} icon={renderAvatar} ref={ref} />;
  }
);

Object.assign(AvatarMenuItem, {
  isMenuChild: true,
  isSelectable: true
});

export default AvatarMenuItem;
