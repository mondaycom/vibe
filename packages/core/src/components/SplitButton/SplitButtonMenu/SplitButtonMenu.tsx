import React, { forwardRef, useRef } from "react";
import useMergeRef from "../../../hooks/useMergeRef";
import Menu, { MenuProps } from "../../Menu/Menu/Menu";

export type SplitButtonMenuProps = Omit<MenuProps, "focusItemIndexOnMount"> & { children: React.ReactNode };

const SplitButtonMenu = forwardRef(
  ({ children, ...splitButtonMenuProps }: SplitButtonMenuProps, ref: React.ForwardedRef<HTMLElement>) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRef(ref, componentRef);

    return (
      <Menu focusItemIndexOnMount={0} {...splitButtonMenuProps} ref={mergedRef}>
        {children}
      </Menu>
    );
  }
);

export default SplitButtonMenu;
