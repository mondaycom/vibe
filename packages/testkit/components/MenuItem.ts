import { Page, Locator, test } from "@playwright/test";
import { BaseElement } from "./BaseElement";

/**
 * Class representing a MenuItem element.
 * Extends the BaseElement class.
 */
export class MenuItem extends BaseElement {
  /**
   * Create a MenuItem.
   * @param {Page} page - The Playwright page object.
   * @param {Locator} locator - The locator for the ListItem element.
   * @param {string} elementReportName - The name for reporting purposes.
   */
  constructor(page: Page, locator: Locator, elementReportName: string) {
    super(page, locator, elementReportName);
  }

  /**
   * Click the MenuItem element.
   * @returns {Promise<void>}
   */
  async click(): Promise<void> {
    await test.step(`Click ${this.elementReportName}`, async () => {
      await this.locator.hover();
      await this.locator.click();
    });
  }
}
