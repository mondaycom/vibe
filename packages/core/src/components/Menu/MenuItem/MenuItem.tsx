import React, { AriaAttributes, ForwardedRef, ReactElement, forwardRef, useMemo, useRef } from "react";
import Tooltip, { TooltipProps } from "../../../components/Tooltip/Tooltip";
import Icon from "../../../components/Icon/Icon";
import useIsOverflowing from "../../../hooks/useIsOverflowing/useIsOverflowing";
import { SubIcon, VibeComponent, VibeComponentProps, withStaticProps } from "../../../types";
import { IconType } from "../../Icon";
import { CloseMenuOption, MenuChild } from "../Menu/MenuConstants";
import Label from "../../Label/Label";
import styles from "./MenuItem.module.scss";
import BaseMenuItem from "./components/BaseMenuItem/BaseMenuItem";
import MenuItemIcon from "./components/MenuItemIcon/MenuItemIcon";
import { TooltipPositions } from "../../Tooltip/Tooltip.types";
import { TooltipPositions as TooltipPositionsEnum } from "../../Tooltip/TooltipConstants";
import { SubmenuPosition } from "./MenuItem.types";

export interface MenuItemProps extends VibeComponentProps {
  /**
   * The title of the menu item.
   */
  title?: string;
  /**
   * The label displayed alongside the title.
   */
  label?: string | React.ReactElement<typeof Label>;
  /**
   * The icon displayed in the menu item.
   */
  icon?: SubIcon;
  /**
   * The type of icon.
   */
  iconType?: IconType;
  /**
   * The background color of the icon.
   */
  iconBackgroundColor?: string;
  /**
   * If true, the menu item is disabled.
   */
  disabled?: boolean;
  /**
   * The reason for disabling the item, shown in a tooltip.
   */
  disableReason?: string;
  /**
   * If true, the menu item is selected.
   */
  selected?: boolean;
  /**
   * Callback fired when the menu item is clicked.
   */
  onClick?: (event: React.MouseEvent | React.KeyboardEvent) => void;
  /**
   * The active item index in the menu.
   */
  activeItemIndex?: number;
  /**
   * Callback to set the active item index.
   */
  setActiveItemIndex?: (index: number) => void;
  /**
   * The index of the menu item.
   */
  index?: number;
  /**
   * The key of the menu item.
   */
  key?: string;
  /**
   * If true, the parent menu is visible.
   */
  isParentMenuVisible?: boolean;
  /**
   * Callback to reset the open submenu index.
   */
  resetOpenSubMenuIndex?: () => void;
  /**
   * If true, a submenu is open.
   */
  hasOpenSubMenu?: boolean;
  /**
   * Callback to open or close a submenu by index.
   */
  setSubMenuIsOpenByIndex?: (index: number, isOpen: boolean) => void;
  /**
   * If true, document event listeners are used for handling interactions.
   */
  useDocumentEventListeners?: boolean;
  /**
   * The tooltip content for the menu item.
   */
  tooltipContent?: string;
  /**
   * The position of the tooltip.
   */
  tooltipPosition?: TooltipPositions;
  /**
   * The delay in milliseconds before the tooltip shows.
   */
  tooltipShowDelay?: number;
  /**
   * Additional props for customizing the tooltip.
   */
  tooltipProps?: Partial<TooltipProps>;
  /**
   * Callback fired when the mouse leaves the item.
   */
  onMouseLeave?: (event: React.MouseEvent) => void;
  /**
   * Callback fired when the mouse enters the item.
   */
  onMouseEnter?: (event: React.MouseEvent) => void;
  /**
   * Class name applied to the icon wrapper.
   */
  iconWrapperClassName?: string;
  /**
   * If true, the menu item starts as selected.
   */
  isInitialSelectedState?: boolean;
  /**
   * If true, the menu scrolls to ensure visibility.
   */
  shouldScrollMenu?: boolean;
  /**
   * Function to close the menu with a given option.
   */
  closeMenu?: (option: CloseMenuOption) => void;
  /**
   * Reference to the menu container.
   */
  menuRef?: React.RefObject<HTMLElement>;
  /**
   * The submenu items, if applicable.
   */
  children?: MenuChild | MenuChild[];
  /**
   * If true, enables a split menu item interaction where the main area triggers an action,
   * while the icon button opens the submenu.
   */
  splitMenuItem?: boolean;
  /**
   * The label of the menu item for accessibility.
   */
  "aria-label"?: AriaAttributes["aria-label"];
  /**
   * The position of a submenu relative to the menu item.
   */
  submenuPosition?: SubmenuPosition;
  /**
   * If true, automatically repositions the submenu when its content changes.
   */
  autoAdjustOnSubMenuContentResize?: boolean;
}

export interface MenuItemTitleComponentProps extends Omit<MenuItemProps, "title"> {
  title: ReactElement;
  "aria-label": NonNullable<AriaAttributes["aria-label"]>;
}

const MenuItem: VibeComponent<MenuItemProps | MenuItemTitleComponentProps> & {
  iconType?: typeof Icon.type;
  isSelectable?: boolean;
  isMenuChild?: boolean;
  tooltipPositions?: typeof TooltipPositionsEnum;
} = forwardRef(
  (
    {
      className,
      iconWrapperClassName,
      title = "",
      label = "",
      icon = "",
      iconType,
      iconBackgroundColor,
      disabled = false,
      disableReason,
      selected = false,
      key,
      children,
      tooltipContent,
      tooltipPosition = "right",
      tooltipShowDelay = 300,
      tooltipProps,
      "aria-label": ariaLabel,
      ...baseMenuProps
    }: MenuItemProps | MenuItemTitleComponentProps,
    ref: ForwardedRef<HTMLElement>
  ) => {
    const titleRef = useRef();

    // if "title" is a component ariaLabel is mandatory
    const iconLabel = ariaLabel ?? (title as string);

    const isTitleHoveredAndOverflowing = useIsOverflowing({ ref: titleRef });
    const shouldShowTooltip = isTitleHoveredAndOverflowing || disabled || tooltipContent;

    const finalTooltipContent = useMemo(() => {
      if (disabled) return disableReason;
      if (tooltipContent) return tooltipContent;
      return title;
    }, [disableReason, disabled, title, tooltipContent]);

    const renderLabel = useMemo(() => {
      if (!label) return;
      if (typeof label === "string") {
        return <Label kind="line" text={label} />;
      }
      if (React.isValidElement(label) && label.type === Label) {
        return label;
      }
    }, [label]);

    return (
      <Tooltip
        content={shouldShowTooltip ? finalTooltipContent : null}
        position={tooltipPosition}
        showDelay={tooltipShowDelay}
        {...tooltipProps}
      >
        <BaseMenuItem
          key={key}
          ref={ref}
          subMenu={children}
          className={className}
          disabled={disabled}
          selected={selected}
          {...baseMenuProps}
        >
          {Boolean(icon) && (
            <MenuItemIcon
              icon={icon}
              type={iconType}
              label={iconLabel}
              disabled={disabled}
              selected={selected}
              backgroundColor={iconBackgroundColor}
              wrapperClassName={iconWrapperClassName}
            />
          )}
          <div ref={titleRef} className={styles.title}>
            {title}
          </div>
          {renderLabel}
        </BaseMenuItem>
      </Tooltip>
    );
  }
);

Object.assign(MenuItem, {
  isSelectable: true,
  isMenuChild: true
});

export default withStaticProps(MenuItem, {
  iconType: Icon.type,
  tooltipPositions: TooltipPositionsEnum
});
