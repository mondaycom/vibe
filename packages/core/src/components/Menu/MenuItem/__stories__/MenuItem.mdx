import { Meta } from "@storybook/blocks";
import { ComponentRuleWithLabelDo, ComponentRuleWithLabelDont } from "./MenuItem.stories.helpers";
import * as MenuItemStories from "./MenuItem.stories";

<Meta of={MenuItemStories} />

# Menu Item

Use menu item for drawing one options that displayed inside a menu.

<Canvas of={MenuItemStories.Overview} />

### Import

```js
import { MenuItem } from "@vibe/core";
```

## Props

<PropsTable />

## Accessibility

<UsageGuidelines
  guidelines={[
    <>
      Use the <code>ariaLabel</code> prop when you need to provide a custom accessible name that differs from the
      visible title text, especially for icon-only menu items.
    </>,
    <>
      For menu items with submenus, ensure the submenu content is properly structured with descriptive menu item titles
      for clear navigation hierarchy.
    </>,
    <>
      Use the <code>disabled</code> prop appropriately with a <code>disableReason</code> to provide context for why a
      menu item cannot be selected.
    </>
  ]}
/>

## Variants

### States

<Canvas of={MenuItemStories.States} />

### Icons

<Canvas of={MenuItemStories.Icons} />

### Label

<Canvas of={MenuItemStories.Label} />

### SubMenu

<Canvas of={MenuItemStories.SubMenu} />

### Overflow

<Canvas of={MenuItemStories.Overflow} />

### Tooltips

<Canvas of={MenuItemStories.TooltipStory} />

## Do's and Donâ€™ts

<ComponentRules
  rules={[
    {
      positive: {
        component: ComponentRuleWithLabelDo(),
        description: "Keep the label text in the menu item short and informative. Use 1 word only."
      },
      negative: {
        component: ComponentRuleWithLabelDont(),
        description: "Dont use long text with more than 1 word, to ensure clarity and to avoid ellipsis."
      }
    }
  ]}
/>
