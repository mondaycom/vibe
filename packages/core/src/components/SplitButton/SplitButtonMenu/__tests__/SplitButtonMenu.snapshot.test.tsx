import { beforeEach, afterEach, describe, it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import SplitButtonMenu from "../SplitButtonMenu";
import MenuItem from "../../../Menu/MenuItem/MenuItem";
import { mockRequestAnimationFrame, restoreRequestAnimationFrameMock } from "../../../../tests/__tests__/test-utils";

const SplitButtonMenuWithItems = (
  <SplitButtonMenu id="menu">
    <MenuItem title="Test 1" />
    <MenuItem title="Test 2" />
  </SplitButtonMenu>
);

describe("SplitButtonMenu", () => {
  beforeEach(() => {
    mockRequestAnimationFrame();
  });

  afterEach(() => {
    restoreRequestAnimationFrameMock();
  });

  it("renders correctly", () => {
    const tree = renderer.create(SplitButtonMenuWithItems).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
