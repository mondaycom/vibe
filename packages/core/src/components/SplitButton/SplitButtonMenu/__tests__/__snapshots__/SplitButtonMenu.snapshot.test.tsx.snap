// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplitButtonMenu renders correctly 1`] = `
<ul
  aria-label="Menu"
  className="menu medium"
  data-testid="menu_menu"
  id="menu"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOver={[Function]}
  role="menu"
  tabIndex={0}
>
  <li
    aria-current={true}
    className="typography primary start singleLineEllipsis text text2Normal item focused initialSelected"
    data-testid="menu-item_0"
    id="menu-item-0"
    onClick={[Function]}
    role="menuitem"
    tabIndex={-1}
  >
    <div
      className="title"
    >
      Test 1
    </div>
  </li>
  <li
    aria-current={false}
    className="typography primary start singleLineEllipsis text text2Normal item initialSelected"
    data-testid="menu-item_1"
    id="menu-item-1"
    onClick={[Function]}
    role="menuitem"
    tabIndex={-1}
  >
    <div
      className="title"
    >
      Test 2
    </div>
  </li>
</ul>
`;
